### ISSUE TITLE
boost: update dependents to conform to composable variants

### ISSUE THREAD

**Issue by trws** [LLNL/spack#25017]
We've been doing some work on the side for this as part of the BUILD project.  Using boost, and this branch, as a build suite we got some data on all the boost components touched by a c++ compiler as part of building each of the packages depending on boost (at least the ones that actually built).  As a result, we have a full list of boost components used by about half the packages touched in here so we can replace the defaults.  Thanks to @ryandomalley  for compiling the data, we have a listing of everything here: 
[Boost.md](https://github.com/spack/spack/files/6857357/Boost.md)

_Originally posted by @trws in https://github.com/spack/spack/issues/22303#issuecomment-884317429_

The full tracking list for this tracking issue is below, if you take one, check the top level box, if you have a PR ready, check the "ready" box and if possible link to it in an edit here.  If you can't add a comment and I'll link it up:

- [ ] 3dtk
    - [ ] ready
- [ ] abyss
    - [ ] ready
- [ ] acts
    - [ ] ready
- [ ] adept-utils
    - [ ] ready
- [ ] apex
    - [ ] ready
- [ ] arrayfire
    - [ ] ready
- [ ] aspcud
    - [ ] ready
- [ ] automaded
    - [ ] ready
- [ ] biobloom
    - [ ] ready
- [ ] bridger
    - [ ] ready
- [ ] cantera
    - [ ] ready
- [ ] canu
    - [ ] ready
- [ ] casper
    - [ ] ready
- [ ] cbtf
    - [ ] ready
- [ ] cbtf-argonavis
    - [ ] ready
- [ ] cbtf-krell
    - [ ] ready
- [ ] cgal
    - [ ] ready
- [ ] clfft
    - [ ] ready
- [ ] coin3d
    - [ ] ready
- [ ] cpprestsdk
    - [ ] ready
- [ ] damaris
    - [ ] ready
- [ ] delly2
    - [ ] ready
- [ ] denovogear
    - [ ] ready
- [ ] dimemas
    - [ ] ready
- [ ] dssp
    - [ ] ready
- [x] dyninst
    - [ ] ready
- [ ] dysco
    - [ ] ready
- [ ] erne
    - [ ] ready
- [ ] express
    - [ ] ready
- [ ] faodel
    - [ ] ready
- [ ] flann
    - [ ] ready
- [ ] flux-sched
    - [ ] ready
- [ ] folly
    - [ ] ready
- [ ] gearshifft
    - [ ] ready
- [ ] gnuradio
    - [ ] ready
- [ ] gource
    - [ ] ready
- [ ] heaptrack
    - [ ] ready
- [ ] helics
    - [ ] ready
- [ ] hpctoolkit
    - [ ] ready
- [x] hpx
    - [ ] ready
- [ ] hssp
    - [ ] ready
- [ ] hyperscan
    - [ ] ready
- [ ] iq-tree
    - [ ] ready
- [ ] isaac-server
    - [ ] ready
- [ ] jali
    - [ ] ready
- [ ] kea
    - [ ] ready
- [x] launchmon
    - [ ] ready
- [ ] libint
    - [ ] ready
- [ ] libkml
    - [ ] ready
- [ ] liblas
    - [ ] ready
- [ ] libpulsar
    - [ ] ready
- [ ] lordec
    - [ ] ready
- [ ] magics
    - [ ] ready
- [ ] mapnik
    - [ ] ready
- [ ] masurca
    - [ ] ready
- [ ] mgis
    - [ ] ready
- [ ] miopen-opencl
    - [ ] ready
- [ ] modern-wheel
    - [ ] ready
- [ ] multiverso
    - [ ] ready
- [ ] muster
    - [ ] ready
- [ ] mysql
    - [ ] ready
- [ ] ncbi-toolkit
    - [ ] ready
- [ ] nektar
    - [ ] ready
- [ ] nix
    - [ ] ready
- [ ] openfoam
    - [ ] ready
- [ ] openimageio
    - [ ] ready
- [ ] openspeedshop
    - [ ] ready
- [ ] pagmo
    - [ ] ready
- [ ] pajeng
    - [ ] ready
- [ ] parsplice
    - [ ] ready
- [ ] pbbam
    - [ ] ready
- [ ] percept
    - [ ] ready
- [ ] povray
    - [ ] ready
- [ ] precice
    - [ ] ready
- [ ] py-espresso
    - [ ] ready
- [ ] py-espressopp
    - [ ] ready
- [ ] py-pycuda
    - [ ] ready
- [ ] py-quast
    - [ ] ready
- [ ] qmcpack
    - [ ] ready
- [ ] revbayes
    - [ ] ready
- [ ] rose
    - [ ] ready
- [ ] sailfish
    - [ ] ready
- [ ] salmon
    - [ ] ready
- [ ] scallop
    - [ ] ready
- [ ] seqan
    - [ ] ready
- [ ] shapeit4
    - [ ] ready
- [ ] shark
    - [ ] ready
- [ ] simgrid
    - [ ] ready
- [ ] source-highlight
    - [ ] ready
- [ ] stat
    - [ ] ready
- [ ] strelka
    - [ ] ready
- [ ] sympol
    - [ ] ready
- [ ] templight-tools
    - [ ] ready
- [ ] tfel
    - [ ] ready
- [ ] thrift
    - [ ] ready
- [ ] tophat
    - [ ] ready
- [ ] trilinos
    - [ ] ready
- [ ] valgrind
    - [ ] ready
- [ ] votca-csg
    - [ ] ready
- [ ] votca-csgapps
    - [ ] ready
- [ ] votca-tools
    - [ ] ready
- [ ] wt
    - [ ] ready
- [ ] xdmf3
    - [ ] ready
- [ ] xios
    - [ ] ready

(edited to sort list)

**Comment by trws** [comment]
@spack/maintainers, if anyone has experience with one of these packages or a minute to set up a PR to update #22303 with the deps for one of these, it would be great.  There are lots of updates to be made here.

**Comment by trws** [comment]
@SteVwonder, any chance you could look at the flux-shed package? I'm pretty sure the list we have in here is accurate, but it would be great to have some active testing on it if we could.

**Comment by trws** [comment]
@ryandomalley, would you be willing to do the folly update?

**Comment by mwkrentel** [comment]
@trws You want updates to the diff in the form of a PR, is that right?
Which repo, branch?

**Comment by ryandomalley** [comment]
> @ryandomalley, would you be willing to do the folly update?

Will do!

**Comment by marvinbernhardt** [comment]
given that #28623 is now merged: is there a (grep?) command or similar to use to find out which boost libraries are needed given the source files?

**Comment by marvinbernhardt** [comment]
For example something like this?
`grep -he 'boost::[a-zA-Z_]*::' **/*.{c,cc,cpp,h} | cut -d ':' -f 3 | sort | uniq`
But for example in votca it will find things like `algorithm` or `bitmaps` and those are not in the list of variants?!

**Comment by hainest** [comment]
>But for example in votca it will find things like algorithm or bitmaps and those are not in the list of variants?!

Variants represent components that require compilation/installation as opposed to being header-only. The composable model makes header-only components available by default. It should be noted that Boost deprecates/removes components and changes which ones require compilation over time. As such, it is critically important to either freeze your Boost version or frequently test your spack package with the latest Boost.

---
P.S. `sort | uniq` can be spelled `sort -u`. :)

**Comment by marvinbernhardt** [comment]
Thanks for clarification.

How is it, will soon many users of the above packages find that they are missing a shared boost library at run time?

**Comment by hainest** [comment]
The packages that depend on boost but do not specify any variants have been updated to use 'Boost.with_default_values'. For example,

    depends_on('boost')
    depends_on(Boost.with_default_values)

I am not in favor of this approach, but it has been merged into develop as of #28959. With this change, packages that `depends_on('boost')` will continue to work without modification. This should be viewed as a short-term fix, and I strongly recommend being explicit in which components your package relies on.

**Comment by trws** [comment]
Yeah, none of us are happy about using the with default values stuff, and have put significant work into detecting variants for a lot of packages (there is another or coming with more). That said, there are quite a few that didn't give us actionable results, because they either just didn't build or wouldn't build with detected subsets or similar. The goal is to get all of this fully specified.
